File: backend/app/models/stream.py
"""WebSocket event and command models — mirrors ``frontend/src/types/stream.ts``.

Events are a discriminated union on ``type``. Each event serialises through
``CamelModel.wire()`` before being pushed to the socket.
"""

from __future__ import annotations

from typing import Dict, List, Optional

from pydantic import Field
from typing_extensions import Literal

from .base import CamelModel
from .memory import (
    MemoryFragment,
    MemoryNarrative,
    ReconstructionStatus,
    SceneParameters,
    SessionSource,
)
from .neural import NeuralFrame
from .pipeline import PipelineStageId, PipelineStageState


class SessionOpenedEvent(CamelModel):
    type: Literal["session.opened"] = "session.opened"
    session_id: str
    sample_rate: int
    channels: List[str]
    duration: Optional[float] = None
    source: SessionSource


class FrameEvent(CamelModel):
    type: Literal["frame"] = "frame"
    frame: NeuralFrame


class PipelineEvent(CamelModel):
    type: Literal["pipeline"] = "pipeline"
    stages: List[PipelineStageState]


class SceneEvent(CamelModel):
    type: Literal["scene"] = "scene"
    scene: SceneParameters


class SceneUpdateEvent(CamelModel):
    type: Literal["scene.update"] = "scene.update"
    region_confidence: Dict[str, float] = Field(default_factory=dict)
    resolved_regions: List[str] = Field(default_factory=list)
    patch: Dict[str, object] = Field(default_factory=dict)


class FragmentEvent(CamelModel):
    type: Literal["fragment"] = "fragment"
    fragment: MemoryFragment


class ReconstructionEvent(CamelModel):
    type: Literal["reconstruction"] = "reconstruction"
    status: ReconstructionStatus
    progress: float
    confidence: float
    coherence: float


class NarrativeEvent(CamelModel):
    type: Literal["narrative"] = "narrative"
    narrative: MemoryNarrative


class SessionClosedEvent(CamelModel):
    type: Literal["session.closed"] = "session.closed"
    session_id: str
    reason: Literal["complete", "stopped", "error"]
    message: Optional[str] = None


class StreamErrorEvent(CamelModel):
    type: Literal["error"] = "error"
    stage: Optional[PipelineStageId] = None
    message: str
    recoverable: bool = True


class StreamCommand(CamelModel):
    """Inbound control message. Unknown fields are ignored by design."""

    type: Literal["start", "pause", "resume", "stop", "seek", "speed", "focus"]
    speed: Optional[float] = None
    t: Optional[float] = None
    region_id: Optional[str] = None

File: frontend/src/types/stream.ts
/**
 * WebSocket protocol between the session console and the backend.
 *
 * The backend pushes a discriminated union of events over `/ws/session/{id}`;
 * the client sends a small set of control commands back on the same socket.
 */

import type { NeuralFrame } from './neural'
import type { PipelineStageId, PipelineStageState } from './pipeline'
import type {
  MemoryFragment,
  MemoryNarrative,
  ReconstructionStatus,
  SceneParameters,
} from './memory'

export interface SessionOpenedEvent {
  type: 'session.opened'
  sessionId: string
  sampleRate: number
  channels: string[]
  /** Total session length in seconds; null for open-ended live streams. */
  duration: number | null
  source: 'simulated' | 'uploaded' | 'device'
}

export interface FrameEvent {
  type: 'frame'
  frame: NeuralFrame
}

export interface PipelineEvent {
  type: 'pipeline'
  stages: PipelineStageState[]
}

/** Emitted once the context model first commits to a scene. */
export interface SceneEvent {
  type: 'scene'
  scene: SceneParameters
}

/** Incremental refinement of an existing scene as evidence accumulates. */
export interface SceneUpdateEvent {
  type: 'scene.update'
  regionConfidence: Record<string, number>
  /** Regions promoted from fragmented to resolved this tick. */
  resolvedRegions: string[]
  /** Parameter drift as the interpretation sharpens. */
  patch: Partial<SceneParameters>
}

export interface FragmentEvent {
  type: 'fragment'
  fragment: MemoryFragment
}

export interface ReconstructionEvent {
  type: 'reconstruction'
  status: ReconstructionStatus
  progress: number
  confidence: number
  coherence: number
}

export interface NarrativeEvent {
  type: 'narrative'
  narrative: MemoryNarrative
}

export interface SessionClosedEvent {
  type: 'session.closed'
  sessionId: string
  reason: 'complete' | 'stopped' | 'error'
  message: string | null
}

export interface StreamErrorEvent {
  type: 'error'
  stage: PipelineStageId | null
  message: string
  recoverable: boolean
}

export type StreamEvent =
  | SessionOpenedEvent
  | FrameEvent
  | PipelineEvent
  | SceneEvent
  | SceneUpdateEvent
  | FragmentEvent
  | ReconstructionEvent
  | NarrativeEvent
  | SessionClosedEvent
  | StreamErrorEvent

export type StreamEventType = StreamEvent['type']

/* -------------------------------------------------------------------------- */
/* Client → server                                                            */
/* -------------------------------------------------------------------------- */

export interface StartCommand {
  type: 'start'
  /** Playback rate multiplier, 0.25–4. */
  speed: number
}

export interface PauseCommand {
  type: 'pause'
}

export interface ResumeCommand {
  type: 'resume'
}

export interface StopCommand {
  type: 'stop'
}

export interface SeekCommand {
  type: 'seek'
  /** Target time in seconds from session start. */
  t: number
}

export interface SpeedCommand {
  type: 'speed'
  speed: number
}

/** Ask the engine to spend extra evidence resolving one region. */
export interface FocusCommand {
  type: 'focus'
  regionId: string
}

export type StreamCommand =
  | StartCommand
  | PauseCommand
  | ResumeCommand
  | StopCommand
  | SeekCommand
  | SpeedCommand
  | FocusCommand

export type ConnectionStatus =
  | 'idle'
  | 'connecting'
  | 'open'
  | 'reconnecting'
  | 'closed'
  | 'failed'
